FIX Add one to current colspan in fixColumns instead of resetting to 2

Kevinn1109 opened the pull request:
<!--
  Thanks for contributing, you're awesome! ⭐

  Please read https://docs.silverstripe.org/en/contributing/code/ if you haven't contributed to this project recently.
-->
## Description
In the old situation, the final column in each row would be hardcoded to colspan="2" if the function 'fixColumns' was called. This is required because this function adds an additional column to the table that needs to be filled in. This change removes the hardcoded number and instead adds one to the current colspan of the last cell. This fixes the bug that the 'No items found' row would be capped at two columns even if the table has more.

## Manual testing steps
1. Manually create a GridField for an has_many relation that contains no records, this should fullfill the check in needsColumnFix by default. The relation class should have at least 3 summary fields.
2. Without this change, the 'No items found' row gets a colspan of 2. With this change, it should be 4.

## Issues
<!--
  List all issues here that this pull request fixes/resolves.
  If there is no issue already, create a new one! You must link your pull request to at least one issue.
-->
- #1859 

## Pull request checklist
<!--
  PLEASE check each of these to ensure you have done everything you need to do!
  If there's something in this list you need help with, please ask so that we can assist you.
-->
- [x] The target branch is correct
    - See [picking the right version](https://docs.silverstripe.org/en/contributing/code/#picking-the-right-version)
- [x] All commits are relevant to the purpose of the PR (e.g. no debug statements, unrelated refactoring, or arbitrary linting)
    - Small amounts of additional linting are usually okay, but if it makes it hard to concentrate on the relevant changes, ask for the unrelated changes to be reverted, and submitted as a separate PR.
- [x] The commit messages follow our [commit message guidelines](https://docs.silverstripe.org/en/contributing/code/#commit-messages)
- [x] The PR follows our [contribution guidelines](https://docs.silverstripe.org/en/contributing/code/)
- [x] Code changes follow our [coding conventions](https://docs.silverstripe.org/en/contributing/coding_conventions/)
- [x] This change is covered with tests (or tests aren't necessary for this change)
- [x] Any relevant User Help/Developer documentation is updated; for impactful changes, information is added to the changelog for the intended release
- [ ] CI is green

ChloeSartorelli commented on client/src/legacy/GridField.js:
```diff
@@ -176,7 +176,8 @@ $.entwine('ss', function($) {
       this.find('tbody tr').each(function () {
         var cell = $(this).find('td:last');
-        cell.attr('colspan', 2);
+        var colspan = cell.attr("colspan") ?? 1;
```
```suggestion
        var colspan = cell.attr('colspan') ?? 1;
```
What is the `?? 1` there for? What scenarios result in there not being a `colspan` attribute?

Kevinn1109 replied:
The colspan attribute is not set by default, I've not seen it in use other than in the case of an empty gridfield.

ChloeSartorelli replied:
Would colspan `2` always be correct when `colspan` wasn't already defined?

ChloeSartorelli commented on client/src/legacy/GridField.js:
```diff
@@ -176,7 +176,8 @@ $.entwine('ss', function($) {
         var cell = $(this).find('td:last');
-        cell.attr('colspan', 2);
+        var colspan = cell.attr("colspan") ?? 1;
+        cell.attr('colspan', colspan + 1);
```
I assume the `+1` is what this is referring to from the PR description:

> This is required because this function adds an additional column to the table that needs to be filled in.

Can you please add a comment about that to the code? As is it's not immediately clear just by glancing at the code why we're doing this.

Kevinn1109 replied:
It's not clear to me either why this still exists. It comes from the SS3 era, and by the look of it the reason for this fix is no longer relevant. Removing it altogether might be the way to go, but that comes with a larger impact. For now I have chosen to only fix the bug that this code causes to ensure that only the context of this bug is affected.

GuySartorelli reviewed (changes requested):
Thanks! Couple of small changes

ChloeSartorelli commented on client/src/legacy/GridField.js:
```diff
@@ -176,7 +177,8 @@ $.entwine('ss', function($) {
       this.find('tbody tr').each(function () {
         var cell = $(this).find('td:last');
-        cell.attr('colspan', 2);
+        var colspan = cell.attr('colspan') ?? 1;
```
```suggestion
        // Note we need to add 1 to the current column span, because we're going to add a new column
        var colspan = cell.attr('colspan') ?? 1;
```

Kevinn1109 committed:
FIX Add one to current colspan in fixColumns instead of resetting to 2

In the previous situation, the 'No items found' row would be hardcapped at colspan=2. This change ensures that that row always spans all columns in the table

ChloeSartorelli commented on client/src/legacy/GridField.js:
```diff
@@ -12,6 +12,7 @@ import '../../../thirdparty/jquery-entwine/jquery.entwine.js';
 $.entwine('ss', function($) {
   $('.grid-field').entwine({
     onmatch: function () {
+      // Adds an actions column for the search action if it does not exist already
```
Please remove this comment from here and add it where this actually happens - probably on the line just before `    fixColumns: function (visible) {` would be perfect.

ChloeSartorelli commented on client/src/legacy/GridField.js:
```diff
@@ -175,8 +176,10 @@ $.entwine('ss', function($) {
     fixColumns: function (visible) {
       this.find('.sortable-header').append('<th class="main col-Actions" />');
       this.find('tbody tr').each(function () {
+        // Note we need to add 1 to the current column span, because we're going to add a new column
```
Please move this comment down one as per the original suggestion in https://github.com/silverstripe/silverstripe-admin/pull/1869#discussion_r1887574735

GuySartorelli reviewed (changes requested):
https://github.com/silverstripe/silverstripe-admin/pull/1869#discussion_r1887577258 hasn't been responded to. We're assuming if the colspan wasn't defined, it is already 1 and should become 2. Does that always hold true? Can we do something like count the number of columns in the table or something?

Kevinn1109 commented:
> [#1869 (comment)](https://github.com/silverstripe/silverstripe-admin/pull/1869#discussion_r1887577258) hasn't been responded to. We're assuming if the colspan wasn't defined, it is already 1 and should become 2. Does that always hold true? Can we do something like count the number of columns in the table or something?

The calling function adds one column to the table, so a correctly set up table will always have rows with the number of columns minus that new column. My change changes a hardcoded 2 that disregards the existing colspan to something that takes it into account. Counting the number of columns isn't needed, the entire purpose of this function is to add one column. If this leads to a wrong number of columns, there's a near certainty that said table already has a wrong number of columns in the current release. As far as I can tell, the only native integration that's affected by this is the 'No items found' row.

In my eyes this is legacy code that should be removed altogether, but I'm not entirely sure if that comes with complications or not.

ChloeSartorelli commented on client/src/legacy/GridField.js:
```diff
@@ -172,11 +172,14 @@ $.entwine('ss', function($) {
-        cell.attr('colspan', 2);
+        // Note we need to add 1 to the current column span, because we're going to add a new column
+        var colspan = cell.attr('colspan') ?? 1;
+        cell.attr('colspan', colspan + 1);
```
In my local testing, I found `colspan` is a string, so this concatenates the values into a new string.
e.g. if `colspan` is `'3'`, we end up setting the new value to `31` instead of the expected `4`.

Kevinn1109 replied:
As it turns out, it can be both an int and a string depending on how the cell is created. I've added support for strings so that that isn't an issue anymore.

ChloeSartorelli commented on client/src/legacy/GridField.js:
```diff
@@ -172,11 +172,14 @@ $.entwine('ss', function($) {
-        cell.attr('colspan', 2);
+        // Note we need to add 1 to the current column span, because we're going to add a new column
+        var colspan = cell.attr('colspan') ?? 1;
+        cell.attr('colspan', +colspan + 1);
```
That syntax seems a bit weird - if this is just to cast to number, please use `Number` instead.
```suggestion
        cell.attr('colspan', Number(colspan) + 1);
```

GuySartorelli reviewed (approved):
Looks good, thanks! Just letting CI run and I'll merge if it's green.
